### PR TITLE
R-IAC algo

### PR DESCRIPTION
**Pull request by sebastien-forestier**
R-IAC-Spliting Interest model
Recursive Tree based on scipy.spatial.kdtree
Spliting modes: 
    - random
    - in the middle of the region
    - in the median of the data
    - choose between random splits the one with max difference of progress between the 2 sub regions
Sampling modes:
    - random in a random leaf weighted by leaf's volume
    - greedy: the leaf with max progress
    - epsilon greedy: greedy with prob. (1-eps) and random with prob. eps
    - softmax: choose between leaves which are weighted by progress*volume plus a softmax exploration (with a temperature parameter)

Ploting possibilities projecting 2D of the Tree
Tests provided (python -m explauto.interest_model.riac)

### CODE REVIEW

File: explauto/interest_model/riac.py [hunk 1]
```diff
@@ -0,0 +1,940 @@
...
+        Competence progress on points of given indexes
+        
+        """
+        if self.progress_measure == 'abs_deriv':
+            if len(idxs) <= 1:
+                return 0
+            else:
+                idxs = sorted(idxs)[- self.progress_win_size:]
+                #print "progress_idxs", idxs, self.get_data_c()[idxs]
+                return abs(np.cov(zip(range(len(idxs)), self.get_data_c()[idxs]), rowvar=0)[0, 1])
```
Comment (sebastien-forestier) [inline explauto/interest_model/riac.py]: Progress measure : absolute value of competence derivative (code from discrete_progress)

File: explauto/interest_model/riac.py [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+                 progress_measure, 
+                 sampling_mode):
+        
+        self.conf = conf
+        self.bounds = self.conf.bounds[:, expl_dims]
+        self.competence_measure = competence_measure
+        
+        if progress_win_size >= max_points_per_region:
+            print "WARNING: progress_win_size should be < max_points_per_region"
+
```
Comment (pierre-rouanet) [inline explauto/interest_model/riac.py]: Is this really only a Warning, or should it never happened? In this case, I will recommend instead something like:

```python
if progress_win_size >= max_points_per_region:
    raise ValueError("WARNING: progress_win_size should be < max_points_per_region")
```

File: explauto/interest_model/riac.py [hunk 3]
```diff
@@ -0,0 +1,947 @@
...
+    
+    def progress(self):
+        return self.tree.progress
+    
+    def get_data_x(self):
+        return self.data_x
+    
+    def get_data_c(self):
+        return self.data_c
+
```
Comment (pierre-rouanet) [inline explauto/interest_model/riac.py]: Why not using property instead ?

File: explauto/interest_model/riac.py [hunk 4]
```diff
@@ -0,0 +1,947 @@
...
+            self.data_c = np.array([self.competence_measure(xy, ms)])
+        else:
+            self.data_c = np.append(self.data_c, self.competence_measure(xy, ms)) 
+        self.tree.add(np.shape(self.data_x)[0] - 1)
+
+
+
+
+
+# Based on scipy.spatial.KDTree
```
Comment (pierre-rouanet) [inline explauto/interest_model/riac.py]: I'm not quite sure, but I'm thinking that instead of copy/pasting the code from the scipy KDTree it should be feasible to use inheritance (or delegation?) instead.

File: explauto/interest_model/riac.py [hunk 5]
```diff
@@ -0,0 +1,947 @@
...
+        """
+        Sample a point in the region of this node
+        
+        """
+        s = rand_bounds(self.bounds_x).flatten()
+        return s
+    
+        
+    def sample(self, sampling_mode = None):
+        """
```
Comment (pierre-rouanet) [inline explauto/interest_model/riac.py]: As you have quite a lot of sampling_mode it would be nice to gather them in some sort of struct (or at least to have the possibilities in the docstring).

Maybe using an [enum](https://docs.python.org/3/library/enum.html)? Maybe an overkill here?

File: explauto/interest_model/riac.py [hunk 6]
```diff
@@ -0,0 +1,947 @@
...
+                progress_max = max(progresses)
+                probas = np.exp(progresses / (progress_max*temperature))
+                probas = probas / np.sum(probas)
+                
+                #print "Softmax", progresses, probas
+                leaf = leaves[np.where(np.random.multinomial(1, probas) == 1)[0][0]]
+                return leaf.sample_bounds()
+            
+        else:
+            raise NotImplementedError(sampling_mode)
```
Comment (pierre-rouanet) [inline explauto/interest_model/riac.py]: This method is actually quite hard to follow. Maybe a solution would be to split them into methods for each sampling mode. And then the sample method could just be something like:
```python

def sample(sampling_mode=None):
    return self.sample_methods[samplig_mode]()
```
